cli/fathom_web/commands/vectorize.py

```diff
@@ -0,0 +1,312 @@
+    seconds seems adequate since one second has always worked for me (Daniel).
+    """
+    for _ in range(5):
+        prefs = (pathlib.Path(firefox.capabilities.get('moz:profile')) / 'prefs.js').read_text().split(';')
+        uuids = next((line for line in prefs if 'extensions.webextensions.uuids' in line)).split(',')
+        try:
+            fathom_fox_uuid = next((line for line in uuids if 'fathomfox@mozilla.com' in line)).split('\\"')[3]
+            return fathom_fox_uuid
+        except StopIteration:
+            time.sleep(1)
+    else:
+        raise GracefulError('Could not find UUID for FathomFox. No entry in the pres.js file.')
+
+
+def vector_files_present(firefox):
+    """Return the number of vector-y files present in Firefox's download dir."""
+    download_dir = pathlib.Path(firefox.profile.default_preferences['browser.download.dir'])
+    vector_files = download_dir.glob('vector*.json')
+    return len(list(vector_files))
+
+
+def extract_error_from(status_text):
+    """
+    Look for errors in the vectorizer's status text.
+
+    If there is an error, raise an exception causing an ungraceful shutdown.
+    """
+    lines = status_text.splitlines()
+    for line in lines:
+        if 'failed:' in line:
+            return line
+    raise UngracefulError(f'There was a vectorizer error, but we could not find it in {status_text}')
+
+
+def look_for_new_vector_file(downloads_dir, vector_files_before):
+    """
+    Look for a new vector file in the downloads directory.
+
+    We use a loop to try multiple times because the file system needs a little
+    little time to update before the file appears. Five seconds seems adequate
+    since one second has always worked for me (Daniel).
+    """
+    for _ in range(5):
+        vector_files_after = set(downloads_dir.glob('vector*.json'))
+        try:
+            new_file = (vector_files_after - vector_files_before).pop()
+            return new_file
+        except KeyError:
+            time.sleep(1)
+    else:
+        raise GracefulError(f'Could not find vectors file in {downloads_dir.as_posix()}.\n{vector_files_before} were already present.')
+
+
+def teardown(firefox, firefox_pid, geckoview_pid, server, server_thread, graceful_shutdown):
+    """
+    Close Firefox and the HTTP server.
+
+    There is a graceful shutdown path and an ungraceful shutdown path. If the
+    vectorizer has started running, we use the ungraceful shutdown path. This
+    is because Firefox becomes unresponsive to the call to `quit()`.
+    """
+    if graceful_shutdown:
+        firefox.quit()
+        server.shutdown()
+        server_thread.join()
+    else:
+        # This is the only way I could get killing the program with ctrl+c to work properly.
+        if server:
+            server.shutdown()
+            server_thread.join()
+        if firefox:
+            try:
+                os.kill(firefox_pid, signal.CTRL_C_EVENT)
```

Comment:
I ran `fathom-vectorize` with the browser open to try and do some debugging, and there was no easy way to pause or stop the vectorization, so I closed the second window which triggered a [couple of errors](https://paste.mozilla.org/DWLYqJjb) -- one related to Selenium Webdriver: `selenium.common.exceptions.StaleElementReferenceException: Message: The element reference of <ul id="status"> is stale; either the element is no longer attached to the DOM, it is not in the current frame context, or the document has been refreshed`, and the other related to calling `teardown` in the `finally` block in `main`: `AttributeError: module 'signal' has no attribute 'CTRL_C_EVENT'`. I'm not sure about the first error, but the second definitely seems worth addressing.

Closing this window also terminated the HTTP server, so I couldn't manually load pages to debug. Maybe we should wait to shut down the HTTP server until after `firefox` has closed/been terminated?

Reply:
I have not found a way to gracefully stop and close the Firefox windows while the vectorization is happening.

How had you paused or stopped the vectorizer before to do debugging?

If the user closes one of the windows, I think the entire vectorization should stop. What do you think?

I'll work on fixing the signal. I didn't realize until now that `CTRL_C_EVENT` was Windows only.

I'll also see if I can switch the order of the HTTP server closing and firefox closing. I recall needed to have this order before, but that reason may no longer be valid. Why are you trying to manually load pages during the vectorization?

Reply:
> How had you paused or stopped the vectorizer before to do debugging?

Typically in FathomFox, I just close the second window, and this would stop vectorization.

> If the user closes one of the windows, I think the entire vectorization should stop. What do you think?

Sounds like a good call, since one window has the pages we're vectorizing, and the other has the status of vectorization. If either closes, we can't really know that we successfully completed vectorization.

> Why are you trying to manually load pages during the vectorization?

I'm not trying to manually load pages during vectorization. I'm trying to pause vectorization and manually load pages in the browser between vectorization attempts.

Right now, FathomFox's `vectors.js` page lists the status, and for some errors, stops vectorization if it encounters an error. Some of these errors are tied to a particular page (e.g. `null` vectors will tell you which page(s) had `null` feature vector values), so maybe I want to open that page up and inspect it to fix the error.

One thing that helps (me) debug such errors (and helps with ruleset development in general) is to amend the `trainees.js` script, so that it actually runs the ruleset against the page as soon as its loaded into the page and logs the results. In this situation, I am using my fathom-trainees/fathom-fox vectorizer setup to load these pages outside of vectorization.

Reply:
> I have not found a way to gracefully stop and close the Firefox windows while the vectorization is happening.

Note: We discussed this in IRC last Friday, when we learned that `driver.quit()` does not work due to this [Firefox bug](https://bugzilla.mozilla.org/show_bug.cgi?id=1430064).